### PR TITLE
Update ArticlePicker to serve pressed articles

### PR DESCRIPTION
**Pull request by ioannakok**
## What does this change?

* It updates `ArticlePicker` to be able to serve pressed articles.
* Renames `PressedInteractives.scala` to PressedContent.scala
* Extracts `decideTier` method and adds `ArticlePickerTest`

## Does this change need to be reproduced in dotcom-rendering ?

- [x] No
- [ ] Yes (please indicate your plans for DCR Implementation)

## What is the value of this and can you measure success?
It delivers the framework for being able to serve pressed articles, which is required to be able to meet OKR "100% of Articles".

When someone presses an article, they will be able to serve the pressed content as opposed to content rendered by DCR.

### Does this affect other platforms?

- [ ] AMP 
- [ ] Apps
- [ ] Other (please specify)

### Does this affect GLabs Paid Content Pages? Should it have support for Paid Content?

- [x] No
- [ ] Yes (please give details)

### Does this change break ad-free?

- [x] No
- [ ] It did, but tests caught it and I fixed it
- [ ] It did, but there was no test coverage so I added that then fixed it

### Does this change update the version of CAPI we're using?

- [x] No, all the existing database files are just fine
- [ ] Yes, and I have [re-run all the tests locally and checked in all the updated data/database/xyz files](https://github.com/guardian/frontend/blob/main/docs/03-dev-howtos/15-updating-test-database.md)

### Accessibility test checklist

- [ ] [Tested with screen reader](https://accessibility.gutools.co.uk/testing/web/screen-readers/)
- [ ] [Navigable with keyboard](https://accessibility.gutools.co.uk/testing/web/keyboard-navigation/)
- [ ] [Colour contrast passed](https://accessibility.gutools.co.uk/testing/web/colour-contrast/)

### Tested

- [x] Locally
- [x] On CODE (optional)

### CODE REVIEW

File: article/app/controllers/ArticleController.scala [hunk 1]
```diff
@@ -144,4 +146,11 @@ class ArticleController(
     }
   }
 
+  def servePressedPage(path: String)(implicit request: RequestHeader): Future[Result] = {
+    val cacheable = WithoutRevalidationResult(
+      Ok.withHeaders("X-Accel-Redirect" -> s"/s3-archive/www.theguardian.com/$path"),
```
Comment (DavidLawes) [inline article/app/controllers/ArticleController.scala:151]: We serve a redirect here because that was the advice we got when implementing equivalent logic for the interactive migration (if I understood correctly, serving the redirect is quicker?).

Note that this redirect is for the prod s3 bucket (https://github.com/guardian/platform/blob/main/router/files/router.conf#L671). We did wonder if we should configure the redirect to work for CODE (i.e. if stage is CODE then point to equivalent CODE s3 bucket). However, there is likely an upcoming improvement we'd like to make where:
- try to get doc from s3
- if success --> serve
- if failure --> render via dcr (even if it's broken, but that's better than serving a 500?)

Comment (OllysCoding) [reply article/app/controllers/ArticleController.scala:151]: If this is the logic for interactives I'd feel comfortable doing the same for articles, but the suggestion for in case is does fail sounds worth looking into!

File: article/app/controllers/ArticleController.scala [hunk 2]
```diff
@@ -109,9 +109,11 @@ class ArticleController(
         Future.successful(common.renderJson(getJson(article), article))
       case EmailFormat =>
         Future.successful(common.renderEmail(ArticleEmailHtmlPage.html(article), article))
+      case HtmlFormat | AmpFormat if tier == PressedArticle =>
+        servePressedPage(path)
       case AmpFormat if isAmpSupported =>
         remoteRenderer.getAMPArticle(ws, article, blocks, pageType)
-      case HtmlFormat | AmpFormat if tier == RemoteRender =>
+      case HtmlFormat | AmpFormat if tier == DotcomRendering =>
```
Comment (oliverlloyd) [inline article/app/controllers/ArticleController.scala:116]: This rename doesn't sync with the function `remoteRenderer`. Can we either lean into it and make the change in all places we use 'remote' or pull back and keep the remote nomenclature. Maybe in a different PR?

File: article/app/services/dotcomponents/ArticlePicker.scala [hunk 3]
```diff
@@ -78,14 +80,13 @@ object ArticlePicker {
...
+      request: RequestHeader,
+  ): RenderType = {
     val checks = dcrChecks(page, request)
     val dcrCanRender = checks.values.forall(identity)
 
-    val tier =
-      if (request.forceDCROff) LocalRenderArticle
-      else if (request.forceDCR || dcrCanRender) RemoteRender
-      else LocalRenderArticle
+    val tier: RenderType = calculateTier(isPressed(ensureStartingForwardSlash(path)), dcrCanRender)
```
Comment (oliverlloyd) [inline article/app/services/dotcomponents/ArticlePicker.scala:89]: `decideTier`?

'calculate' implies math to me

File: article/app/services/dotcomponents/ArticlePicker.scala [hunk 4]
```diff
@@ -96,12 +97,27 @@ object ArticlePicker {
       ("dcrCouldRender" -> dcrCanRender.toString) +
       ("pageTones" -> pageTones)
 
-    if (tier == RemoteRender) {
+    if (tier == DotcomRendering) {
       DotcomponentsLogger.logger.logRequest(s"path executing in dotcomponents", features, page)
+    } else if (tier == PressedArticle) {
+      DotcomponentsLogger.logger.logRequest(s"Article served from pressed content", features, page)
```
Comment (oliverlloyd) [inline article/app/services/dotcomponents/ArticlePicker.scala:103]: I'm not entirely sure here but it might be safer to add metadata to record this information rather than change this balance?

It's hard to say. I can see how what you've done makes a lot of sense in terms of the fact there are now 3 places the code can go to render the article. On the other hand, this text is used to produce a number of graphs that provide information around DCR/Frontend traffic patterns and this change would impact them.

On yet another hand, are we using those graphs anymore?

In any case, if you did decide to go with this approach - and I'm happy to leave this choice up to you - is there value in keeping the text consistent? Eg. less `Article served` more `path executing`.

File: article/app/services/dotcomponents/ArticlePicker.scala (hunk 3 above)
Comment (oliverlloyd) [reply article/app/services/dotcomponents/ArticlePicker.scala:89]: Also, what's the value of this abstraction. Could we just keep this code inline?

File: article/app/services/dotcomponents/RenderType.scala [hunk 5]
```diff
@@ -1,7 +1,8 @@
 package services.dotcomponents
 
 sealed trait RenderType
-case object RemoteRender extends RenderType
-case object RemoteRenderAMP extends RenderType
-case object LocalRenderAmp extends RenderType
-case object LocalRenderArticle extends RenderType
+case object DotcomRendering extends RenderType
+case object DotcomRenderingAMP extends RenderType
+case object FrontendLegacyAMP extends RenderType
```
Comment (oliverlloyd) [inline article/app/services/dotcomponents/RenderType.scala:6]: Can we move this rename into a separate PR, please? This is a _monumental_ change and probably shouldn't be an aside but celebrated on it's own! 😎

**Review by oliverlloyd — COMMENTED** [review]
Amazing work!

File: article/app/services/dotcomponents/ArticlePicker.scala (hunk 3 above)
Comment (ioannakok) [reply article/app/services/dotcomponents/ArticlePicker.scala:89]: Yes, `decideTier` makes more sense indeed! We'll change that. As for the abstraction, it enabled us to easily write unit tests for this logic. Otherwise we would have to mock `PageWithStoryPackage`. This abstraction was suggested by Roberto as an easy / good way of handling the situation.

Comment (oliverlloyd) [reply article/app/services/dotcomponents/ArticlePicker.scala:89]: So, and this is totally an aside, I have noticed that this approach tends to lead to a codebase that is heavily abstracted with lots of unit tests. Some people like this and I'm not saying it is wrong but, personally, I find this creates a situation where the code is hard to maintain. Changes or edits to code require tests to be updated and involve switching context between files and having to check to make sure you're not impacting other places in the code that might be importing the abstraction.

My own, personal preference is to favour integration tests where the interface that your testing is the same as the readers (or other consumers of the code like routes or apis requests) and then to keep code inline where it's closer to where it's being used and easier to maintain.

In general, Frontend seems like a unit test heavy codebase where abstractions are encouraged and so arguably it's right that this pattern be followed here. Maybe this type of code is better suited to it? My _point_ is, please don't take this as a rule and use it everywhere!

File: article/app/services/dotcomponents/ArticlePicker.scala (hunk 4 above)
Comment (ioannakok) [reply article/app/services/dotcomponents/ArticlePicker.scala:103]: We will have to update the dashboards to reflect the percentage of pressed content (as a follow on task). After the migration to 100% Articles, we are expecting to have 0 content rendered from frontend. We expect the frontend tier to act as a fallback and all the content to be rendered by either DCR or pressed. 

This feels like an intermediate step until the frontend code gets deleted. After that we will only have 2 rendering tiers and at that point we will have to refactor both the logging and the dashboards again.

File: article/app/controllers/ArticleController.scala (hunk 2 above)
Comment (ioannakok) [reply article/app/controllers/ArticleController.scala]: Makes sense, we reverted the renaming for now.

File: article/app/services/dotcomponents/RenderType.scala (hunk 5 above)
Comment (ioannakok) [reply article/app/services/dotcomponents/RenderType.scala]: Done!

File: article/app/controllers/ArticleController.scala [hunk 6]
```diff
@@ -109,6 +109,8 @@ class ArticleController(
         Future.successful(common.renderJson(getJson(article), article))
       case EmailFormat =>
         Future.successful(common.renderEmail(ArticleEmailHtmlPage.html(article), article))
+      case HtmlFormat | AmpFormat if tier == PressedArticle =>
+        servePressedPage(path)
```
Comment (oliverlloyd) [inline article/app/controllers/ArticleController.scala:113]: This is really nice 👍

File: article/app/services/dotcomponents/ArticlePicker.scala [hunk 7]
```diff
@@ -98,10 +99,25 @@ object ArticlePicker {
...
     }
 
     tier
   }
+
+  def decideTier(isPressed: Boolean, dcrCanRender: Boolean)(implicit
+      request: RequestHeader,
+  ): RenderType = {
+    if (request.forceDCROff) LocalRenderArticle
+    else if (isPressed) PressedArticle
```
Comment (oliverlloyd) [inline article/app/services/dotcomponents/ArticlePicker.scala:115]: Question: If I pass a url for a pressed article with the flag `?dcr=false` should I still get the Pressed version or the Frontend version?

Comment (oliverlloyd) [reply article/app/services/dotcomponents/ArticlePicker.scala:115]: If we still want to serve pressed content regardless the flag then we should maybe swap these lines?

```suggestion
    if (isPressed) PressedArticle
    else if (request.forceDCROff) LocalRenderArticle
```

File: article/app/services/dotcomponents/ArticlePicker.scala [hunk 8]
```diff
@@ -98,10 +99,25 @@ object ArticlePicker {
...
+    else if (request.forceDCR || dcrCanRender) RemoteRender
```
Comment (oliverlloyd) [inline article/app/services/dotcomponents/ArticlePicker.scala:116]: We also need to consider the semantics of the request saying `dcr=true` which is saying 'force this request to go to DCR' - it might be useful and more accurate if we respect this for pressed content?

File: common/app/services/dotcomrendering/PressedContent.scala [hunk 9]
```diff
@@ -1,14 +1,14 @@
...
+  // Temporarily retain a list of pressed content
   // Ed Tools are supporting us in how to batch/programmatically add tags to pressed articles.
   // When we can tag pressed articles (tracking/dcroptout) then we will:
   // - tag all articles that appear in this list
   // - remove this file entirely
-  // - update the InteractiveController to show pressed pages based on presence of the tag
-  // - update the press+clean functionality to automate tagging as part of this process
-  private[this] val interactives = Set[String](
+  // - update the ArticleController and the InteractiveController to show pressed pages based on presence of the tag
+  // - update the press+clean functionality to automate tagging as part of this process?
```
Comment (oliverlloyd) [inline common/app/services/dotcomrendering/PressedContent.scala:10]: I love how this addition of a question mark carries so much weight 😄

**Review by oliverlloyd — APPROVED** [review]
Happy to approve but I think we might need to consider the logic around that if statement. I think there are a couple of new paths we're introducing that need capturing. I _think_ it goes like this but you should check

```js
if (force DCR) DCR // No matter what
else if (pressed) Pressed // Pressed content is always served from S3
else if (force NOT DCR) Frontend // Allow Frontend to be bypassed
else if (dcrCanRender) DCR // This is the normal article path
else Frontend // This is (was) the failover, not much ends up here now!
```

File: article/app/services/dotcomponents/ArticlePicker.scala (hunk 7 above)
Comment (DavidLawes) [reply article/app/services/dotcomponents/ArticlePicker.scala:115]: If passing the query param `?dcr-false` we should get the frontend rendered version. We want forceDCROff to be prioritised over pressing in the following situations:
- we want to validate how the article renders in frontend, e.g. why was this article pressed and are you sure the pressed version matches what we previously had in frontend?
- we want/need to re-press the article, e.g. the article owner updated the page content and we need to re-press.

I'd suggest leaving the order in this case.

Comment (oliverlloyd) [reply article/app/services/dotcomponents/ArticlePicker.scala:115]: _Ideally_ there would be a `?pressed=false` flag we could use...

**Review by oliverlloyd — APPROVED** [review]
🚢

**Review by jamesgorrie — APPROVED** [review]
🥳
